File: Excel/500-599/505/505 Challenge.py
import pandas as pd
import string

def create_triangular_dataframe(letters):
    def triangular_number(n):
        return (n * (n + 1)) // 2
    
    n = max([i for i in range(1, 100) if triangular_number(i) <= len(letters)])
    num_rows = n + 1
    num_cols = n * 2 + 1
    df = pd.DataFrame("", index=range(num_rows), columns=range(num_cols))
    letters_idx = 0
    
    def fill_row(row_num, num_letters):
        nonlocal letters_idx
        start_col = (num_cols - (2 * num_letters - 1)) // 2
        for j in range(num_letters):
            df.iat[row_num, start_col + j * 2] = letters[letters_idx]
            letters_idx += 1
    
    for i in range(1, n + 1):
        fill_row(i - 1, i)
    
    fill_row(num_rows - 1, len(letters) - triangular_number(n))
    print(df)

# Example usage
create_triangular_dataframe(list(string.ascii_uppercase))

File: Excel/500-599/505/505 Challenge.R
library(dplyr)
library(purrr)

create_triangular_dataframe <- function(letters) {
  n <- max(which((1:10 * (1:10 + 1)) / 2 <= length(letters)))
  num_rows <- n + 1
  num_cols <- n * 2 + 1
  
  df <- data.frame(matrix("", nrow = num_rows, ncol = num_cols))
  
  letters_idx <- 1
  
  fill_row <- function(row_num, num_letters) {
    start_col <- (num_cols - (2 * num_letters - 1)) %/% 2 + 1
    df[row_num, seq(start_col, by = 2, length.out = num_letters)] <<- letters[letters_idx:(letters_idx + num_letters - 1)]
    letters_idx <<- letters_idx + num_letters
  }
  
  walk(1:n, ~ fill_row(.x, .x))
  
  fill_row(num_rows, length(letters) - (n * (n + 1)) / 2)

  colnames(df) <- LETTERS[1:ncol(df)]  
  print(df, row.names = FALSE)
}

df = create_triangular_dataframe(LETTERS)
